ddt.py

```diff
@@ -18,59 +20,398 @@
 # They are added to the decorated test method and processed later
 # by the `ddt` class decorator.
 
-DATA_ATTR = '%values'      # store the data the test must run with
-FILE_ATTR = '%file_path'   # store the path to JSON file
-UNPACK_ATTR = '%unpack'    # remember that we have to unpack values
+PARAMS_SETS_ATTR = '%params_sets'   # store a list of *ParamsSet objects
+UNPACKALL_ATTR = '%unpackall'       # remember @unpackall decorators
+
+
+# Public interface - Decorators
 
 
 def unpack(func):
     """
-    Method decorator to add unpack feature.
+    Method decorator to unpack parameters from the next (syntactically)
+    parameters set (``@data`` or ``@file_data`` decorator) by one level.
+
+    Multiple levels are unpacked if ``@unpack`` and ``@unpackall`` are combined
+    and/or applied multiple times.
+
+    """
+    getattr(func, PARAMS_SETS_ATTR)[0].unpack()
+    return func
+
+
+def unpackall(func):
+    """
+    Method decorator to unpack parameters in all parameter sets by one level.
+
+    Multiple levels are unpacked if ``@unpack`` and ``@unpackall`` are combined
+    and/or applied multiple times.
 
     """
-    setattr(func, UNPACK_ATTR, True)
+    setattr(func, UNPACKALL_ATTR, getattr(func, UNPACKALL_ATTR, 0) + 1)
     return func
 
 
-def data(*values):
+def data(*unnamed_values, **named_values):
     """
-    Method decorator to add to your test methods.
+    Method decorator to supply parameters to your test methods in-line.
 
     Should be added to methods of instances of ``unittest.TestCase``.
 
+    Keyword arguments can be used to explicitly define names of values to be
+    used in names of created test methods.
+
     """
     def wrapper(func):
-        setattr(func, DATA_ATTR, values)
+        if not hasattr(func, PARAMS_SETS_ATTR):
+            setattr(func, PARAMS_SETS_ATTR, [])
+        params = InlineParamsSet(*unnamed_values, **named_values)
+        getattr(func, PARAMS_SETS_ATTR).insert(0, params)
         return func
     return wrapper
 
 
-def file_data(value):
+def file_data(filepath, encoding=None):
     """
-    Method decorator to add to your test methods.
+    Method decorator to supply parameters to your test method from a JSON file.
```

Comment:
Can I suggest that while you're changing this, you move to allow YAML data? It will parse json, but then also allow for a more readable style of data as well.